### PR TITLE
Track intervention UUID in classification metadata for downstream clients

### PR DESCRIPTION
**Pull request by camallen**
Track the intervention payload UUID in the next classification metadata, this links the any classification after an intervention message was seen.

linked to https://github.com/zooniverse/interventions-gateway/issues/28

Staging branch URL:

Fixes # .

Describe your changes.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

File: app/redux/ducks/classify.js [hunk 1]
```diff
@@ -264,8 +286,9 @@ export function addIntervention(data) {
 }
 
 export function clearIntervention() {
-  return {
-    type: CLEAR_INTERVENTION
+  return (dispatch) => {
+    dispatch({type: STORE_INTERVENTION_UUID});
```
Comment (camallen) [inline app/redux/ducks/classify.js]: I wasn't sure that this was the way to do it, this code path isn't tested in specs by the look. Thoughts appreciated.

File: app/redux/ducks/classify.js [hunk 2]
```diff
@@ -173,11 +187,19 @@ export default function reducer(state = initialState, action = {}) {
       const { goldStandardMode } = state;
       const { project } = action.payload;
       const { workflow } = state;
+      let { lastInterventionUUID} = state;
```
Comment (camallen) [inline app/redux/ducks/classify.js]: I didn't take into account if the user is actually opted into the interventions and thus saw the message. I think this variable needs to be reflect this state before propagating to the `createNewClassification` method. Thoughts?

File: app/redux/ducks/classify.spec.js [hunk 3]
```diff
@@ -278,6 +307,26 @@ describe('Classifier actions', function () {
...
+      expect(newState.classification.metadata.hasOwnProperty('intervention_uuid')).to.be.false;
+    });
+    describe('with lastInteventionUUID set', function () {
+      const interventionUUIDState = {
+        classification: { id: '1' },
+        workflow: { id: '1' },
+        upcomingSubjects: [mockSubject('1')],
+        lastInterventionUUID: '2d931510-d99f-494a-8c67-87feb05e1594'
+      };
+      it('should record the lastInteventionUUID as metadata.intervention_uuid', function () {
```
Comment (srallen) [inline app/redux/ducks/classify.spec.js]: Typo in the it statement. It should be `should record the lastInterventionUUID as metadata.intervention_uuid`

File: app/redux/ducks/classify.js (hunk 2 above)
Comment (srallen) [reply app/redux/ducks/classify.js]: If I'm understanding this question correctly, there are two pieces of information you need to know? Am I also correct in understanding this to mean that there can be a last intervention UUID, but the user may not have actually seen the intervention? If that's the case, then perhaps the state can be an object and the UUID and seen state is explicitly stored. Something like?: `{ lastIntervention: { UUID: '', seen: boolean } }` and this also gets recorded in the classification metadata?

File: app/redux/ducks/classify.js (hunk 1 above)
Comment (srallen) [reply app/redux/ducks/classify.js]: As someone who isn't familiar with the implementation of interventions in general, reading this is confusing and hard to follow: A clear intervention function is called but then is calling an action to store the id in the state, then in the classification metadata, then the id gets cleared from the state? I would suggest at the very least adding a separate redux action for clearing the UUID if there's going to be an action for storing it. Calling to store the UUID here doesn't seem quite like the right place to me either. The naming of `clearIntervention` would not convey to me that actions are called in the function that do store data.

Comment (eatyourgreens) [reply app/redux/ducks/classify.js]: Firing two state changes together should be ok, but I don’t know if Redux guarantees that the new state after saving the UUID will be the initial state for clearing the intervention. 

I’d be inclined to return a single action which remembers the UUID then clears the intervention, so state is only updated once.

File: app/redux/ducks/classify.js (hunk 2 above)
Comment (eatyourgreens) [reply app/redux/ducks/classify.js]: If you call actions from the `Intervention` component, I think that guarantees they will only be called for volunteers who have opted to see interventions.

File: app/redux/ducks/classify.js (hunk 1 above)
Comment (eatyourgreens) [reply app/redux/ducks/classify.js]: The specs for these actions are the first tests I wrote for redux in PFE and I'm not sure any more that testing the actions directly is right.

You should be able to test the `clearIntervention()` action creator if you import it and the reducer into the tests, and pass it a mock `dispatch`. Have a look at the specs for the translations action creators, where I used that approach. The mock dispatch function I came up with is this one:
https://github.com/zooniverse/Panoptes-Front-End/blob/9461eb8f83b4b12e0b9bee501cbe8d535783e1a7/app/redux/ducks/translations.spec.js#L26-L31

File: app/redux/ducks/classify.js (hunk 2 above)
Comment (eatyourgreens) [reply app/redux/ducks/classify.js]: Reading through this some more, `lastInterventionUUID` is set when the `Intervention` component unmounts, so I think it can only be truthy here if an intervention was shown to the volunteer.

If I'm understanding these changes correctly:

1. I see an intervention and continue.
2. The intervention UUID is stored and the seen intervention is cleared.
3. When the next subject loads, `createNewClassification` is called, the UUID is added to the new classification and the stored UUID is cleared.

File: app/redux/ducks/classify.js [hunk 4]
```diff
@@ -148,6 +154,14 @@ export default function reducer(state = initialState, action = {}) {
       }
       return state;
     }
+    case STORE_INTERVENTION_UUID: {
+      const { intervention } = state;
+      let lastInterventionUUID = null
+      if (intervention && intervention.hasOwnProperty('uuid')) {
+        lastInterventionUUID = intervention.uuid;
+      }
+      return Object.assign({}, state, { lastInterventionUUID });
```
Comment (eatyourgreens) [inline app/redux/ducks/classify.js]: This returns a new state object even if there's no intervention. Is that intentional? I think it should return the current state in that case, but I'm not sure.

File: app/redux/ducks/classify.js [hunk 5]
```diff
@@ -148,6 +154,14 @@ export default function reducer(state = initialState, action = {}) {
       }
       return state;
     }
+    case STORE_INTERVENTION_UUID: {
+      const { intervention } = state;
+      let lastInterventionUUID = null
+      if (intervention && intervention.hasOwnProperty('uuid')) {
```
Comment (eatyourgreens) [inline app/redux/ducks/classify.js]:
```suggestion
      if (intervention && intervention.uuid) {
```

File: app/redux/ducks/classify.spec.js (hunk 3 above)
Comment (camallen) [reply app/redux/ducks/classify.spec.js]:
```suggestion
      it('should record the lastInterventionUUID as metadata.intervention_uuid', function () {
```

File: app/redux/ducks/classify.js (hunk 1 above)
Comment (eatyourgreens) [reply app/redux/ducks/classify.js]: I've pushed a test for the `clearIntervention` action creator.

**Comment by eatyourgreens** [comment]
I think that would require changing the behaviour of the classifier, so couldn't be done in a refactor. At the moment, the intervention is cleared when you close the Intervention component and finish a subject, then the new classification is created when the next subject loads.

**Comment by srallen** [comment]
I've realized that my interpretation of what's going on was incorrect because of how github is previewing the diff of that file, so you can ignore my suggestion. I'll read it over again, but in VSCode instead.
